api/api-doc/storage_service.json

```diff
@@ -2852,6 +2852,46 @@
             }
          ]
       },
+      {
+         "path":"/storage_service/tablets/repair",
+         "operations":[
+            {
+               "nickname":"repair_tablet",
+               "method":"POST",
+               "summary":"Repair a tablet",
+               "type":"void",
+               "produces":[
+                  "application/json"
+               ],
+               "parameters":[
+                  {
+                     "name":"ks",
+                     "description":"Keyspace name to repair",
+                     "required":true,
+                     "allowMultiple":false,
+                     "type":"string",
+                     "paramType":"query"
+                  },
+                  {
+                     "name":"table",
+                     "description":"Table name to repair",
+                     "required":true,
+                     "allowMultiple":false,
+                     "type":"string",
+                     "paramType":"query"
+                  },
+                  {
+                     "name":"tokens",
+                     "description":"Tokens owned by the tablets to repair. Multiple tokens can be provided using a comma-separated list. When set to the special word 'all', all tablets will be repaired",
```

Comment:
I do not like to overload empty to present the meaning of all. Empty can also mean selecting none. Users can specify the token parameter but provides empty content. It could also be it is just that users forget to provide the token parameter by mistake.  Always providing the tokens explicitly reduces corner cases.

Reply:
I'd still argue that it is more intuitive and convenient like I described. `nodetool repair` already works like this. If you don't specify keyspaces, it repairs all of them. It's common to design APIs like this.

Reply:
nodetool repair calls the restful api that repairs a specific keyspace one after another. Scylla manger always repair a single table at a time.  Repair all keyspaces or tables in a single repair request is not what we use. It is bad because one would have to abort the whole repair request if a single tablet in one tablet fails. Repairing one tablet per request makes much more sense and is what we use in production.

Reply:
> Specifying `token` with empty value is different from not specifying the parameter. The handler can distinguish that.

Yes, but why would we introduce all the corner cases. If the api marks the token parameter as mandatory as we do in this PR,  the user will always provide the tokens parameter. 

Also, the user of the api is management tool not the end user.  Typing one less parameter for convenience is not important at all. Being explicit for the parameters to avoid mistakes sounds more important.

Reply:
> Yes, but why would we introduce all the corner cases.

To make the API easier to use. Repairing everything is a common action, e.g. executed manually by the user.

I'd rather be able to do `curl -X POST http://localhost:10000/storage_service/tablets/repair` without having to type anything more.

> Being explicit for the parameters to avoid mistakes sounds more important.

What kind of mistakes?

Reply:
> To make the API easier to use. Repairing everything is a common action, e.g. executed manually by the user.
> 
> I'd rather be able to do `curl -X POST http://localhost:10000/storage_service/tablets/repair` without having to type anything more.

As I mentioned, this api is a manual repair api. It is designed to be consumed by tools not by users. 

If you want to type less, use `nodetool repair`. If you want do not want to even type `nodetool repair`, use auto repair. 

> What kind of mistakes?

The user intends to repair a specific tablet or tablet, but forgets the parameters, then everything is repaired in busy hours.

Reply:
> The user intends to repair a specific tablet or tablet, but forgets the parameters, then everything is repaired in busy hours.

But you said that "It is designed to be consumed by tools not by users.". Tools will be tested before-hand and correct.

If a user made a mistake and pressed enter too soon, he can cancel the repair using cancel API. I can't imagine that he expects that only token X will be repaired without him typing token X into the command line.

Reply:
The tools could also be the user's own script. I meant something like this in the code. 
The user wants to repair tokens 1,2,3, but forgets to add the tokens parameter in the code.

So instead of 
```
curl -X POST "http://127.0.0.1:10000/storage_service/tablets/repair?ks=ks1&table=standard1&tokens=1,2,3"
```
the following is issued and executed without the user even notices. 
```
curl -X POST "http://127.0.0.1:10000/storage_service/tablets/repair?ks=ks1&table=standard1"
```
If we require the tokens to be mandatory, such errors could be avoided. 

Another example for empty tokens means all tokens
```
tokens = get_tokens_to_repair() // it returns empty tokens means nothing to repair
curl -X POST "http://127.0.0.1:10000/storage_service/tablets/repair?ks=ks1&table=standard1&tokens=${tokens}"
```
The above will repair all tokens incorrectly.

So, typing less in the api request is less important than being clear and avoiding possible errors.

Reply:
> If we require the tokens to be mandatory, such errors could be avoided.

True.

But in this case, the user's mistake is obvious from looking at the URL, it's not like the unintended behavior is surprising. And it is not catastrophic. So ease of use should win. 

The same argument holds for "nodetool repair", and the existing "repair_async" API, which had "ranges" param optional. So it's consistent with existing APIs.

> Another example for empty tokens means all tokens
> 
> tokens = get_tokens_to_repair() // it returns empty tokens means nothing to repair
> curl -X POST "http://127.0.0.1:10000/storage_service/tablets/repair?ks=ks1&table=standard1&tokens=${tokens}"
> The above will repair all tokens incorrectly.
> 
> So, typing less in the api request is less important than being clear and avoiding possible errors.
> 

As for this example, I agree, and this is not what I suggested. Like I mentioned, empty "tokens" should be treated as nothing to repair. empty param should be distinguished from unset param in the restful API handler.

Reply:
I don't think we reached consensus here. @avikivity Please say if you're fine with this.

Reply:
> As for this example, I agree, and this is not what I suggested. Like I mentioned, empty "tokens" should be treated as nothing to repair. empty param should be distinguished from unset param in the restful API handler.

This alone adds lots of confusing. As I mentioned multiple times, being explicit and having less corner is cases is much much more important than allowing people to type a few words in an API that is designed to be used by programs or scripts.